### ISSUE TITLE
[BUG] ScaleBar shows incorrect length

### ISSUE THREAD

**Issue by S-Man42** [fleaflet/flutter_map#1950]
### What is the bug?

I really love to see this feature. However, it seems not to be correct.

The scale shows a distance of 1000km in the horizontal distance.

![Screenshot_20240819_210047_Brave](https://github.com/user-attachments/assets/f141dcb0-4e2d-4652-ad49-6e0d20f0ae83)

In fact, it is not more than 650km. 

I guess it results from the problem, that the mercator projection yield different distances throughout the map for same pixel distances depending on latitude. 

But in this case it is not quite clear which latitude it shows. Does it measures the rhumb line or geodetic lengths? Currently it may show wring results as you can see.

I guess it could be a good idea to configure the scale this way, that it shows the distance of the start and end points lying exactly under the bar. 

Second point would be to decide which length you really display. The shortest length (geodetic) between start and end point of the scale, but in that case it is not a straight line and the half marker would not be at the drawed position. If one would show the length of the straight line as the scale bar shows, you would need the length of the rhumb line. (in mercator projection).

Funny enough: Googles scale is also not that precise, I believe:

![Screenshot_20240821_073423_Maps](https://github.com/user-attachments/assets/91ccd155-c8d2-4bd0-916b-d46c2928fef5)

### How can we reproduce it?

Add ScaleBar layer and scroll to any area you know its diameter

### Do you have a potential solution?

I would tried following, I don't know if it works

(for left alignment, others are similar)

Take the pixel position of the starting point of the scale. Convert it to the coordinate exactly below it. Calculate the rhumb line length exactly <currently. displayed scale unit/meters, i.e. 200km> east. Convert this back to the current pixel coords as end point, draw line.

In my opinion this should ensure that the scale exactly shows the length of the rhumbline which is exactly below it. So one could do measuring by panning the map right under the scale. If would choose rhumbline over geodetic's length because it shows a straight line on a mercator projection and hence it should not calc a "curved" geodetic's length in my opinion. It's relevant for huge distances. For smaller ones one could think about switching to geodetics since the performance could be probably increased and difference is very small.

### Platforms

Tested in web demo

### Severity

Minimum: Allows normal functioning

**Comment by S-Man42** [comment]
Other solution could be offering a vertical scale with two or even three sections (halfs, thirds) which show that then scale change when going north or south.

**Comment by ibrierley** [comment]
Not quite sure if this is possibly related to https://github.com/fleaflet/flutter_map/issues/1314

**Comment by JaffaKetchup** [comment]
I believe it shows the scale at the center of the viewport, so I guess it makes sense that it's out when the center of the viewport is much closer to latitude 0 than Germany at this zoom level - I assume the center of the screen is in the South Atlantic Ocean somewhere.
I guess it could make sense for the scale bar to actually measure the scale where it is, but then it gives a less accurate overview at these low zoom levels.

At higher zoom levels the problem won't be as large. When I place it over a place I use to calibrate  exactly, it still reads correct because at that scale there's probably less than 10km between scale bar and viewport center.

**Comment by josxha** [comment]
I agree that the scale bar can be misleading on smaller zoom levels. However, I don't think much can be done to improve this. (and for example Google Maps on mobile has the same behavior). 
I agree with @JaffaKetchup and I believe the scale bar is a useful tool to give an idea about distances but to have it 100% accurate for all of the visible viewport is not possible with all projections.

**Comment by S-Man42** [comment]
Since the scale bar length depends on the latitude, it could be an interesting feature to add a vertical scale with increasing sectors which show the same actual length but grow on increasing lat. I believe it would be a great and afaik unique feature.
![Screenshot_20240823_082129_Locus Map Classic](https://github.com/user-attachments/assets/c811e92d-583a-41cc-918d-163545d48db0)

**Comment by josxha** [comment]
@S-Man42 this would be a feature not a bug, wouldn't it? There has been some discussion in the past about enhancing the scale bar, like show it in miles or a combined meter/miles scale bar similar to what google maps has. Pull request are welcome.

**Comment by S-Man42** [comment]
Hm, maybe. It could be seen as a feature, I believe. It depends on how you classify this issue. If you say: Well, it is not a bug since it does what it should do in your opinion and the issue is... well... one couldn't do anything. Then it is a new feature. Otherwise one could see it as probable workaround to enhance the usability and go around the described issue ;)

It's just brainstorming. Maybe I'll find some time to add this to the project. :)

**Comment by monsieurtanuki** [comment]
The notion of ScaleBar for low zoom levels makes no sense anyway: what would be the scale on zoom 0?
I would suggest not displaying the ScaleBar for low zoom levels.
If really needed, the more relevant option would be some kind of flat ellipse (e.g. 100km wide), draggable and depending on the latitude.

**Comment by JaffaKetchup** [comment]
I'm going to clarify this as a feature request, as it's impossible to get a single behaviour that will work right for everyone with the current scale bar.

**Comment by S-Man42** [comment]
Created a precise scalebar vor my own map based on this code here. It is not currently for merging due major changes. Maybe one could merge the codes into one using several new options. One has to be extremely careful with pixels and positions. For example: The current code takes a position on the map (center) , calculates the bar length, draws it as a new widget and positions it afterwards with padding. Making the measuring precisely at the drawed line the padding position must be taken into account during calculation. One cannot use the bottom left pixel when the scalebar will be drawn 30 pixels higher instead. These 30px are important for length calculations on low zoom. levels... Such things... 

Take this as a PoC. If I find the time, I will make a more generalized version but of course this can be done by anybody else :) 

https://github.com/S-Man42/GCWizard/tree/3.2.0/lib/tools/coords/map_view/widget/scalebar

![Screenshot_20240831_235751](https://github.com/user-attachments/assets/b96765c2-eae1-42ea-bf45-c7d0f0bb42b6)

**Comment by monsieurtanuki** [comment]
@S-Man42 A scale _grid_ would even make more sense:
* centered on the map center
* showing the scale in the 4 directions NSEW - instead of just North South in your screenshot

**Comment by S-Man42** [comment]
Yes, indeed, it would be a nice feature too. However since we discussed a bar not a grid, I implemented the first one.

**Comment by monsieurtanuki** [comment]
@S-Man42 Fair enough. It's just that the OP was about a horizontal scale bar. Your scale bar is vertical. And I believe the solution is a grid.

**Comment by S-Man42** [comment]
I did this because it demonstrates the increasing scale and therefore the used precision which is not necessary/possible in horizontal orientation

**Comment by josxha** [comment]
I like the idea of enhancing the scale bar in general. But I struggle a bit at the moment to think of a use case where the exact scale at every pixel of the screen is crucial and where a measure-distance-tool won't be used. @S-Man42 could you describe some example use case where this is mandatory please?

**Comment by JaffaKetchup** [comment]
I really like the look of the vertical scalebar, even though I don't know how widely it would be used in general since users are much more used to a horizontal scalebar. If it was improved enough to be resilient as you said, I think it's the sort of thing that maybe we could put in the core, otherwise it would make a great plugin. I'm not going to look into introducing the vertical scalebar right now (partially since your code is under GPLv3) - but thanks for making a working demo!

I'm going to close this issue for now for the reasons above.
